### PR TITLE
avm1: implement scroll, maxscroll and bottomScroll for TextField

### LINKED ISSUE #1033: TextField scroll properties are unimplemented
Flash:
image
Ruffle:
image

Here is the FLA and SWF
[bottomScroll.zip](https://github.com/ruffle-rs/ruffle/files/5111699/bottomScroll.zip)

### PR DESCRIPTION
**Pull request by midgleyc**
Implement scroll, maxscroll and bottomScroll for avm1's TextField.

Closes #1033.

### CODE REVIEW

File: core/src/display_object/edit_text.rs [hunk 1]
```diff
@@ -850,6 +855,101 @@ impl<'gc> EditText<'gc> {
         }
     }
 
+    /// How many lines the text can be scrolled down
+    pub fn maxscroll(self) -> usize {
+        let edit_text = self.0.read();
+
+        // non-word-wrapped text can't be scrolled
```
Comment (adrian17) [inline core/src/display_object/edit_text.rs:862]: I see counterexamples, like:
```
tf.text = "1\n2\n3\n4\n5\n6\n7\n"; // only 4 lines fit
tf.multiline = true;
tf.wordWrap = false;
trace(tf.maxscroll); // 4
```

**Review by adrian17 — COMMENTED** [review]
Did you test it on the scroll-related reports?
Most seem to be nicely fixed, but:
- https://github.com/ruffle-rs/ruffle/issues/4217 - the scrolling text field still doesn't seem to be working
- https://github.com/ruffle-rs/ruffle/issues/4416 - last line doesn't fit well (probably related to font change and not related to your PR, just mentioning it here for completeness)

**Comment by adrian17** [comment]
Also side observation: looks like - at least in FP7 - there's a bug that `maxscroll` doesn't force reflow?

See:
```js
tf.multiline = true;
tf.wordWrap = false;
tf.text = "abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc";
trace(tf.maxscroll);
```
Correct, prints 1.

```js
tf.multiline = true;
tf.text = "abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc";
tf.wordWrap = false;
trace(tf.maxscroll);
```
Incorrect, prints 5. (but correctly displays text in one line)

Same code, with forced text reflow:

```
tf.multiline = true;
tf.text = "abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc abc";
tf.wordWrap = false;
trace(tf.textHeight);
trace(tf.maxscroll);
```
Correctly prints `1`.

Note that I'm just mentioning this as a remark, as this modern-DOM-like "lazy relayout on property read" isn't implemented anyways. Fixing this is definitely out of scope of this PR, so in the tests I recommend just setting `.text` last to avoid issues :)

File: core/src/avm1/globals/text_field.rs [hunk 2]
```diff
@@ -617,3 +620,41 @@ pub fn maxhscroll<'gc>(
...
+}
+
+pub fn set_scroll<'gc>(
+    this: EditText<'gc>,
+    activation: &mut Activation<'_, 'gc, '_>,
+    value: Value<'gc>,
+) -> Result<(), Error<'gc>> {
+    let input = value.coerce_to_f64(activation)?;
+    // the greatest input has a fractional part as well, but I didn't compute it
+    let scroll_lines = if input.is_nan() || input < 0.0 || input >= 767100486418433.0 {
```
[the remark this answers is not in the document]
Comment (midgleyc) [reply core/src/avm1/globals/text_field.rs:638]: I derived it experiementally. It's not precisely correct: the value is somewhere between 767100486418432.9 and 767100486418433. I assume it's to do with their internal representation, but I couldn't figure out what it was.

**Comment by midgleyc** [comment]
For #4217: if I use the desktop version, I can't see the text at all, so I can't test the scroll.

**Comment by adrian17** [comment]
That's probably because the site downloads text separately, in an .xml file. Either download the files and put them next to the .swf (I _think_ that should work), or build an extension and run it on the page.

File: core/src/display_object/edit_text.rs (hunk 1 above)
Comment (midgleyc) [reply core/src/display_object/edit_text.rs]: hah, I failed to test anything with manual line breaks! Good spot!

**Comment by midgleyc** [comment]
Looking at ekg.swf, it looks like the scroll fields are being set correctly. I don't know why the scroll buttons aren't clickable:
```
textField: [object #889] {
              bottomScroll: 3
              maxscroll: 6
              scroll: 1
              maxhscroll: 0
              hscroll: 0
              password: false
              wordWrap: true
              variable: null
              type: "dynamic"
              textWidth: 400
              textHeight: 126
              textColor: 0
              text: "The patient has left. A copy of the ECG is placed on the left side of the desk. You should now put it in the appropriate section of the binder for future reference! There are four sections in the binder with a couple of sample ECG's showing four common conditions. See if you can decide which section of the binder to put your ECG copy in! Move the cursor over a tab to see what a section contains and click to open the binder to that section. To put the ECG in the binder section of your choice, simply select the appropriate condition from the menu at the bottom of the ECG you have just obtained."
              selectable: true
              multiline: true
              length: 599
              htmlText: "The patient has left. A copy of the ECG is placed on the left side of the desk. You should now put it in the appropriate section of the binder for future reference! There are four sections in the binder with a couple of sample ECG's showing four common conditions. See if you can decide which section of the binder to put your ECG copy in! Move the cursor over a tab to see what a section contains and click to open the binder to that section. To put the ECG in the binder section of your choice, simply select the appropriate condition from the menu at the bottom of the ECG you have just obtained."
              html: false
              embedFonts: false
              borderColor: 0
              border: false
              backgroundColor: 16777215
              background: false
              autoSize: "none"
              vScroller: [object #884]
}
```

**Comment by danielhjacobs** [comment]
There are some scroll-related issues on some sites I maintain that I never reported because I figured there were a significant amount of issues about scroll properties already. I thought it would be good to test them with this branch. Here are my results:

* https://dnalc.cshl.edu/view/2022-G2C-3-D-Brain.html (made up of 3 swfs calling each other, check the network tab) - Scrollbar now works. A possibly unrelated issue still exists; not all the select options are viewable and there's no way to scroll the select box down, so you can't choose anything below Amygdala.

* http://www.dnai.org/timeline/, click 1950-54, hover over the right side of the animation until it scrolls to a picture of Dr. Watson, click the picture ([watson_bio.zip](https://github.com/ruffle-rs/ruffle/files/6721985/watson_bio.zip)) - Scrollbar now works.

* http://www.insidecancer.org/, click "Causes and Prevention", click page "3", click "→" ([causes_map.zip](https://github.com/ruffle-rs/ruffle/files/6721990/causes_map.zip)) - Scrollbar still does not work on the lung cancer legend.

**Comment by midgleyc** [comment]
There's an `onScroller` event that's supposed to be invoked when one of the scroll properties changes. It's used by the library for the broken scrollbar. That's probably it.

**Comment by midgleyc** [comment]
`onScroller` seems to fire whenever the text is laid out again (e.g. if wordwrap flips to off, but no scroll properties change). Implementing this the naive way lead to an error (stack overflow on main), so I'll put this off for now.

**Comment by midgleyc** [comment]
More details about `onScroller`, for lack of a better place:

* I was (probably) wrong about it firing on layout. Flipping wordwrap changes `maxhscroll` on type input. On type dynamic, if flipping wordwrap is the first change you make, it sends, but not after that -- this might be related to lazy loading of scroll properties.
* No matter how many properties change, it only fires once.
* If the text is "input", and you input enough to go to another line / go back a line, it fires.

The way I expect it works for the currently broken scroller is that it starts off blank, gets populated by the XML, `onScroller` fires, then the scrollbar is up and working.

**Comment by midgleyc** [comment]
Filed #4792 for ekg.swf and the "Causes and Prevention" part of inside cancer.

The fourth SWF seems to use a different scrollbar.

**Pull request merged**